core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveBroadcastToPartitionedJoin.java

```diff
@@ -0,0 +1,460 @@
+                                    .with(source().matching(exchange()
+                                            .matching(AdaptiveBroadcastToPartitionedJoin::isBroadcastRemoteExchange)
+                                            .capturedAs(RIGHT_EXCHANGE_NODE))))))),
+                    // local exchange + remote exchange pattern
+                    prev -> prev.with(right().matching(exchange().matching(node -> node.getScope() == LOCAL)
+                            .with(source().matching(exchange()
+                                    .matching(AdaptiveBroadcastToPartitionedJoin::isBroadcastRemoteExchange)
+                                    .capturedAs(RIGHT_EXCHANGE_NODE))))),
+                    // remote exchange pattern
+                    prev -> prev.with(right().matching(exchange()
+                            .matching(AdaptiveBroadcastToPartitionedJoin::isBroadcastRemoteExchange)
+                            .capturedAs(RIGHT_EXCHANGE_NODE))),
+                    // remote source node or any arbitrary node
+                    prev -> prev.with(right().matching(any())));
+
+    private final CostComparator costComparator;
+    private final TaskCountEstimator taskCountEstimator;
+
+    private static boolean isRoundRobinRemoteExchange(ExchangeNode node)
+    {
+        return node.getScope() == REMOTE
+                && node.getType() == REPARTITION
+                && node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION)
+                && node.getSources().size() == 1;
+    }
+
+    private static boolean isBroadcastRemoteExchange(ExchangeNode node)
+    {
+        return node.getScope() == REMOTE
+                && node.getType() == REPARTITION
+                && node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION)
+                && node.getSources().size() == 1;
+    }
+
+    public AdaptiveBroadcastToPartitionedJoin(CostComparator costComparator, TaskCountEstimator taskCountEstimator)
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        // This rule is only enabled in case of FTE
+        return getRetryPolicy(session) == TASK
+                && isFaultTolerantExecutionAdaptiveBroadcastToPartitionedJoinEnabled(session);
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        if (mustReplicate(node, context)) {
+            return Result.empty();
+        }
+        boolean isExtraRemoteExchangeNeededAtProbeSide = captures.getOptional(LEFT_EXCHANGE_NODE).isEmpty();
```

Comment:
How does having a remote exchange on left side ensures that we do not need to repartition again.
You are checking if there is an remote exchange with `FIXED_ARBITRARY_DISTRIBUTION`. Where do we guarantee that data is actually distributed according to join keys?

Reply:
Oh - ok the meaning of that is:
* are we changing existing exchange or adding new one.